STMicroelectronics/SPWF04Sx/SPWF04SxInterface.cs

```diff
@@ -187,6 +187,418 @@ public class SPWF04SxInterface : NetworkInterface, ISocketProvider, ISslStreamPr
             return result.Substring(result.IndexOf(':') + 1);
         }
 
+        //*************************************** Begin Commands added by RoSchmi  ***************************************************
+
+
+        /// <summary>
+        /// Computes a hash
+        /// 0:SHA1, 1:SHA224, 2:SHA256, 3:MD5
+        /// </summary>
+        /// <param name="function">The methode to use. 0:SHA1, 1:SHA224, 2:SHA256, 3:MD5</param>
+        /// <param name="filename">The name of the file that holds the data
+        public byte[] ComputeHash(string function, string filename)   // 0:SHA1, 1:SHA224, 2:SHA256, 3:MD5
+        {
+            var cmd = this.GetCommand()
+                .AddParameter(function)
+                .AddParameter(filename)
+                .Finalize(SPWF04SxCommandIds.HASH);
+            this.EnqueueCommand(cmd);
+            byte[] totalBuf = new byte[0];
+            byte[] lastBuf = new byte[0];
+            int offset = 0;
+            byte[] readBuf = new byte[50];
+            int len = readBuf.Length;
+            while (len > 0)
+            {
+                len = cmd.ReadBuffer(readBuf, 0, len);
+                lastBuf = totalBuf;
+                offset = lastBuf.Length;
+                totalBuf = new byte[offset + len];
+                Array.Copy(lastBuf, 0, totalBuf, 0, offset);
+                Array.Copy(readBuf, 0, totalBuf, offset, len);
+                readBuf = new byte[len];
+            }
+            this.FinishCommand(cmd);
+            return totalBuf;
+        }
+
+        public void Reset()
+        {
+            var cmd = this.GetCommand()
+               .Finalize(SPWF04SxCommandIds.RESET);
+            this.EnqueueCommand(cmd);
+            cmd.ReadBuffer();
+            this.FinishCommand(cmd);
+        }
+
+        public void SendAT()
+        {
+            var cmd = this.GetCommand()
+               .Finalize(SPWF04SxCommandIds.AT);
+            this.EnqueueCommand(cmd);
+            cmd.ReadBuffer();
+            this.FinishCommand(cmd);
+        }
+
+        /// <summary>
+        /// Mounts a memory volume
+        /// 0 = External Flash, 1 = User Flash, 2 = Ram, 3 = Application Flash
+        /// </summary>
+        /// /// <param name="volume">The volume to be mounted. 0 = External Flash, 1 = User Flash, 2 = Ram, 3 = Application Flash</param>
+        public void MountMemoryVolume(string volume)
+        {
+            if (!((volume == "0") || (volume == "1") || (volume == "2") || (volume == "3")))
+            { throw new NotSupportedException("Invalid volume. For volume only 0, 1, 2 or 3 are allowed"); }
+            var cmd = this.GetCommand()
+                .AddParameter(volume)
+                .Finalize(SPWF04SxCommandIds.FSM);
+            this.EnqueueCommand(cmd);
+            byte[] readBuf = new byte[50];
```

Comment:
Try not to allocate in commands like this, devices like the FEZ are _very_ memory sensitive. We introduced the object pool abstraction to try and help with fragmentation for this type of thing.